### PR TITLE
[WIP] implementation of whole-grained Petri nets

### PR DESCRIPTION
**Pull request by olynch**
This is a working pull request, not ready to be merged yet. I want to replace the code in AlgebraicPetri with code that uses this new implementation, and also create some visualization code.

### CODE REVIEW

File: src/Petri.jl [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+add_output!(p::AbstractPetri,t,s) = add_part!(p,:O,(os=s,ot=t))
+add_output!(p::AbstractPetri,n,t,s) = add_parts!(p,:O,n,(os=s,ot=t))
+
+# Note: although indexing makes this pretty fast, it is often faster to bulk-convert
+# the Petri net into a transition matrix, if you are working with all of the transitions
+inputs(p::AbstractPetri,t) = subpart(p,incident(p,t,:it),:is)
+outputs(p::AbstractPetri,t) = subpart(p,incident(p,t,:ot),os)
+
+totransmat(p::AbstractPetri) = begin
```
Comment (jpfairbanks) [inline src/Petri.jl:48]: Would it be more idiomatic to call this function `transition_matrix` or make it a constructor of `Array{Int,2}`?

File: src/Petri.jl [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    for i in 1:ni(p)
+        A[subpart(p,i,:it),subpart(p,i,:is),1] += 1
+    end
+    for i in 1:no(p)
+        A[subpart(p,i,:ot),subpart(p,o,:os),2] += 1
+    end
+    A
+end
+
+fromtransmat(A::Array{Int64,3}) = begin
```
Comment (jpfairbanks) [inline src/Petri.jl:59]: Should this be a constructor? ex. `Petri(A::Array{T, 3}) where {T<: Integer}`

File: src/Petri.jl [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            add_outputs!(p,A[i,j,2],i,j)
+        end
+    end
+    p
+end
+
+rate_eq(p::AbstractPetri) = begin
+    n,m = nt(p),ns(p)
+    rates = zeros(m)
+    A = totransmat(p)
```
[the remark this answers is not in the document]
Comment (olynch) [reply src/Petri.jl:76]: Yes.

File: src/Petri.jl (hunk 1 above)
Comment (olynch) [reply src/Petri.jl:48]: Do you mean `Array{Int,3}`? Also, I don't know what would be more idiomatic, and I am not attached to nomenclature; whatever you think is the right way to do name this I will do.

File: src/Petri.jl (hunk 2 above)
Comment (olynch) [reply src/Petri.jl:59]: See my previous reply.

File: src/Petri.jl [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    f(du,u,p,t) = begin
+        for i in 1:n
+            rates[i] = p[i] * prod(u[j] ^ T[i,j,1] for j in 1:m)
```
[the remark this answers is not in the document]
Comment (olynch) [reply src/Petri.jl:79]: Oooh, fancy. OK, will do!

**Comment by olynch** [comment]
Oh, and I believe that the scourge of 4 spaces has once again slipped into the codebase. I will put this nuisance to the sword with a change to my emacs config once and for all!

File: src/Petri.jl (hunk 1 above)
Comment (jpfairbanks) [reply src/Petri.jl:48]: Oh yeah, 3D array makes sense, I got confused by the two in this call `A = zeros(Int64,(nt(p),ns(p),2))` but that is saying the third dimension has length 2. Looking at the code below where this gets used, I think that making this two sparse matrices one for `A[:,:,1]` and the other for `A[:,:,2]` and returning the pair fo them as Tuple{CSCMatrix{Int}, CSCMatrix{Int}}` would be better.

File: src/Petri.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (jpfairbanks) [reply src/Petri.jl:79]: If you do the log transform on the rates and represent `A` as a sparse matrix, you can calculate this with `log.(p) .+ A*log.(u)`.

File: src/Petri.jl [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        end
+        for j in 1:m
+            du[j] = sum(rates[i] * (T[i,j,2] - T[i,j,1]) for i in 1:n)
```
Comment (jpfairbanks) [inline src/Petri.jl:82]: With the pair of matrices representation this becomes `du = (A[2] - A[1])*exp.(rates)`

**Comment by mehalter** [comment]
This is a great start for getting the Petri net implementation moved over to CSets, I agree with all the feedback that James brought up, and think after those changes get made we should get this merged into a branch on the repo. I think we can merge it into a new feature development branch until it is ready to fully replace the initial implementation, but having a feature branch will make it easier for us to collaborate without putting unready code on master

**Pull request merged**